src/PostgREST/App.hs

```diff
@@ -9,430 +9,664 @@ Some of its functionality includes:
+                             QualifiedIdentifier, ReadRequest)
+
+import Protolude      hiding (Handler, toS)
+import Protolude.Conv (toS)
+
+
+data RequestContext =
+  RequestContext
+    { rConfig      :: AppConfig
+    , rDbStructure :: DbStructure
+    , rApiRequest  :: ApiRequest
+    , rContentType :: ContentType
+    }
+
+type Handler = ExceptT Error
+
+type DbHandler = Handler SQL.Transaction
+
+-- | PostgREST application
+postgrest
+  :: Types.LogLevel
+  -> IORef AppConfig
+  -> IORef (Maybe DbStructure)
+  -> SQL.Pool
+  -> IO UTCTime
+  -> IO () -- ^ Lauch connection worker in a separate thread
+  -> Wai.Application
 postgrest logLev refConf refDbStructure pool getTime connWorker =
-  pgrstMiddleware logLev $ \ req respond -> do
-    time <- getTime
-    body <- strictRequestBody req
-    maybeDbStructure <- readIORef refDbStructure
-    conf <- readIORef refConf
-    case maybeDbStructure of
-      Nothing -> respond . errorResponseFor $ ConnectionLostError
-      Just dbStructure -> do
-        response <- do
-          let apiReq = userApiRequest (configDbSchemas conf) (configDbRootSpec conf) dbStructure req body
-          case apiReq of
-            Left err -> return . errorResponseFor $ err
-            Right apiRequest -> do
-              -- The jwt must be checked before touching the db.
-              attempt <- attemptJwtClaims (configJWKS conf) (configJwtAudience conf) (toS $ iJWT apiRequest) time (configJwtRoleClaimKey conf)
-              case jwtClaims attempt of
-                Left errJwt -> return . errorResponseFor $ errJwt
-                Right claims -> do
-                  let
-                    authed = containsRole claims
-                    shouldCommit   = configDbTxAllowOverride conf && iPreferTransaction apiRequest == Just Commit
-                    shouldRollback = configDbTxAllowOverride conf && iPreferTransaction apiRequest == Just Rollback
-                    preferenceApplied
-                      | shouldCommit    = addHeadersIfNotIncluded [(hPreferenceApplied, BS.pack (show Commit))]
-                      | shouldRollback  = addHeadersIfNotIncluded [(hPreferenceApplied, BS.pack (show Rollback))]
-                      | otherwise       = identity
-                    handleReq = do
-                      when (shouldRollback || (configDbTxRollbackAll conf && not shouldCommit)) HT.condemn
-                      mapResponseHeaders preferenceApplied <$> runPgLocals conf claims (app dbStructure conf) apiRequest
-                  dbResp <- P.use pool $ HT.transaction HT.ReadCommitted (txMode apiRequest) handleReq
-                  return $ either (errorResponseFor . PgError authed) identity dbResp
-        -- Launch the connWorker when the connection is down. The postgrest function can respond successfully(with a stale schema cache) before the connWorker is done.
-        when (responseStatus response == status503) connWorker
+  Middleware.pgrstMiddleware logLev $
```

Comment:
It's odd that we're passing in the log level and refConf separately. I think adding the middleware should be done in `Main.hs`. This is not part of the main app - it's just some additional stuff. Also this would remove the whole logging from the spec tests - which makes sense, because it's not used at all.

This might drop coverage by a few lines in `Middleware.hs` temporarily. But we need to add a few logging tests anyway and can do so for the cors stuff again, too.

After that, `\req respond ->` should be part of the main function arguments - reducing nesting/indentation a lot here.

Reply:
Yes, much cleaner.

Reply:
Moving this to Main.hs changes the semantics, several Cors tests failed. Backed out the commit - recommend that we leave this unchanged for this PR.